src/python/zquantum/core/wip/circuits/__init__.py

```diff
@@ -1,4 +1,116 @@
-"""Public API for the ZQuantum circuits."""
+"""Public API for the ZQuantum circuits.
+
+Allows:
+
+- defining quantum circuits with gates applied to qubits
+- set of built-in gates (see imports in this `__init__`)
+- import/export from/to Cirq, PyQuil & Qiskit circuits
+- circuit (de)serialization to/from JSON-compatible dicts
+
+Examples:
+
+    Defining a circuit with a NOT gate on qubit 0 & Hadamard gate on qubit 1::
+        circuit = Circuit()
+        circuit += X(0)
+        circuit += H(1)
+
+    Adding 2-qubit gates::
+        circuit += CNOT(0, 4)
+
+    Adding parametrized gates::
+        circuit += RX(np.pi / 2)(0)
+        circuit += RX(sympy.sympify("theta / 2"))(1)
+
+    Adding 2-qubit parametrized gates::
+        circuit += CPHASE(0.314)(3, 2)
+
+    Adding a built-in gate by its name::
+        circuit += builtin_gate_by_name("X")(0)
+        circuit += builtin_gate_by_name("RX")(np.pi * 1.5)(1)
+
+    Binding parameters::
+        circuit = circuit.bind({sympy.Symbol("theta"): -np.pi / 5})
+
+    Iterating over circuit contents::
+        for gate_op in circuit.operations:
+            name = gate_op.gate.name
+            params = gate_op.gate.params
+            qubits = gate_op.qubit_indices
+            print(f"{name} with params {params} applied to {qubits}")
+
+    Making a different circuit (e.g. shifting gates by 1 qubit):: 
+        new_circuit = Circuit(
+            operations=[
+                gate_op.gate(*[qubit + 1 for qubit in gate_op.qubits])
+                for gate_op in circuit.operations
+            ],
+            n_qubits=circuit.n_qubits
+        )
+
+    Conversion to other frameworks::
+        export_to_cirq(circuit)
+        circuit2 = import_from_cirq(cirq_circuit)
+
+        export_to_pyquil(circuit)
+        circuit3 = import_from_pyquil(pyquil_program)
+
+        export_to_qiskit(circuit)
+        circuit4 = import_from_qiskit(qiskit_quantum_circuit)
+
+    (De)serialization::
+        to_dict(circuit)
+        circuit5 = circuit_from_dict(dict5)
+
+
+Defining new gates
+------------------
+
+To use a gate that isn't already covered by built-in ones you can define a custom gate
+or extend the set of the built-in ones and file a PR to z-quantum-core.
+
+Using custom gates::
+    CustomA = circuits.CustomGateDefinition(
```

Comment:
I would change it to `custom_A` as it's a more pythonic name of the variable.

Reply:
Good point! Updated in 21c28f3.